src/signtool/IBM_Container.cpp

```diff
@@ -0,0 +1,1767 @@
+    m_contFldMap["container-size"]         =  e_CONTAINER_SIZE;
+    m_contFldMap["target-hrmor"]           =  e_TARGET_HRMOR;
+    m_contFldMap["stack-pointer"]          =  e_STACK_POINTER;
+    m_contFldMap["hw-keya"]                =  e_HW_PUBLIC_KEY_A;
+    m_contFldMap["hw-keyb"]                =  e_HW_PUBLIC_KEY_B;
+    m_contFldMap["hw-keyc"]                =  e_HW_PUBLIC_KEY_C;
+
+    m_contFldMap["hdr-version"]            =  e_PRE_HDR_VERSION;
+    m_contFldMap["hdr-hash-algo"]          =  e_PRE_HDR_HASH_ALGORITHM;
+    m_contFldMap["hdr-sign-algo"]          =  e_PRE_HDR_SIGNING_ALGORITHM;
+    m_contFldMap["hdr-code-start-offset"]  =  e_PRE_HDR_CODE_START_OFFSET;
+    m_contFldMap["hdr-flags"]              =  e_PRE_HDR_FLAGS;
+    m_contFldMap["hdr-payload-size"]       =  e_PRE_HDR_PAYLOAD_SIZE;
+    m_contFldMap["hdr-payload-hash"]       =  e_PRE_HDR_PAYLOAD_HASH;
+
+    m_contFldMap["hw-signa"]               =  e_HW_SIGNATURE_A;
+    m_contFldMap["hw-signb"]               =  e_HW_SIGNATURE_B;
+    m_contFldMap["hw-signc"]               =  e_HW_SIGNATURE_C;
+    m_contFldMap["sw-keyp"]                =  e_SW_PUBLIC_KEY_P;
+    m_contFldMap["sw-keyq"]                =  e_SW_PUBLIC_KEY_Q;
+    m_contFldMap["sw-keyr"]                =  e_SW_PUBLIC_KEY_R;
+
+    m_contFldMap["sw-version"]             =  e_SW_HDR_VERSION;
+    m_contFldMap["sw-hash-algo"]           =  e_SW_HDR_HASH_ALGORITHM;
+    m_contFldMap["sw-code-start-offset"]   =  e_SW_HDR_CODE_START_OFFSET;
+    m_contFldMap["sw-flags"]               =  e_SW_HDR_FLAGS;
+    m_contFldMap["sw-payload-size"]        =  e_SW_HDR_PAYLOAD_SIZE;
+    m_contFldMap["sw-payload-hash"]        =  e_SW_HDR_PAYLOAD_HASH;
+
+    m_contFldMap["sw-signp"]               =  e_SW_SIGNATURE_P;
+    m_contFldMap["sw-signq"]               =  e_SW_SIGNATURE_Q;
+    m_contFldMap["sw-signr"]               =  e_SW_SIGNATURE_R;
+
+    m_hdrFldTypeMap["prefix_hdr"]          =  e_FLD_PREFIX_HDR;
+    m_hdrFldTypeMap["software_hdr"]        =  e_FLD_SOFTWARE_HDR;
+
+    m_hashAlgoMap["sha1"]                  =  e_SHA1_ALGO;
+    m_hashAlgoMap["sha256"]                =  e_SHA256_ALGO;
+    m_hashAlgoMap["sha384"]                =  e_SHA384_ALGO;
+    m_hashAlgoMap["sha512"]                =  e_SHA512_ALGO;
+}
+
+
+void IBM_Container::GetFieldNameList( std::vector<std::string>& p_fldNameList )
+{
+    p_fldNameList.clear();
+
+    std::transform( m_contFldMap.begin(), 
+                    m_contFldMap.end(),
+                    std::inserter( p_fldNameList, p_fldNameList.begin() ),
+                    GetKey<ContainerFldMap::value_type>() );
+}
+
+
+// Given a stream of bytes, parse the data and to extract the container fields
+void IBM_Container::ParseContainer( const std::vector<uint8_t>& p_rawData )
+{
+    const uint8_t* pRawData = &p_rawData[0];
+
+    THROW_EXCEPTION( pRawData == NULL );
+
+    int keyCount = 0;
+
+    // Parse the ContainerHdr
+    m_containerHdr.m_magicNumber = getUint32( pRawData );
```

Comment:
I would have done 'getUint32', very different.
```
namespace details
{
    template <typename T> struct getField
    {
        void op(T&, uint8_t**) = delete;
    };

    template <> struct getField<uint8_t>
    {
        void op(uint8_t& value, uint8_t** buffer)
        {
             value = **buffer;
             *buffer += sizeof(value);
        }
    };
    /// ... duplicate for other types, doing endianness fixups as appropriate.
}

template < typename T > void getField(T& value, uint8_t** buffer) { return details::getField<T>::op(value, buffer); }
```

Now all of these lines are just:

    getField(m_containerHdr.m_magicNumber, pRawData);

or if you want to be explicit:

    getField<uint32_t>(m_containerHdr.m_magicNumber, pRawData);